core/src/bson_type_info.rs

```diff
@@ -33,19 +43,27 @@ pub struct BsonTypeInfo {
     // The maximum or actual character length of a character string or binary data type. It is the
     // maximum character length for a fixed-length data type, or the actual character length for a
     // variable-length data type. Its value always excludes the null-termination byte that ends the
-    // character string.
-    pub length: Option<u16>,
+    // character string. This is a function since a BI tool may set a maximum string length. For
+    // most types, the function argument is ignored and a default value is returned; for the String
+    // type, if a max string length is set, it is argued to the function and returned.
+    pub length: fn(Option<u16>) -> Option<u16>,
```

Comment:
I'll explain why I chose this type for `length` and the other updated fields in `BsonTypeInfo`. I considered a few options.

The ticket stated that we wanted to "set the `display_size`, `column_size`, `length`, and `char_octet_length` for `BsonTypeInfo::STRING`" if the max string length option was enabled. Note that we want to do this _only for_ `String`. So, to achieve this, we had a few options:
- In each relevant getter method (`length(&self, type_mode: TypeMode) -> Option<u16>`, etc), do:
```
...
.map(|value| match max_string_length {
    Some(max_string_length) => value.min(max_string_length),
    None => value
})
```
- ^ this would come at the end of each getter, after determining whether to use the `BsonTypeInfo` field or its underlying `SimpleTypeInfo` value.
  - Rejected since it is clunky to essentially match on both the field and the `max_string_length` every time these are called to check that they are both `Some` and then manually calculate the minimum each time. Plus, this would apply to all types, not just `String`, which is wasteful.
- Include a conditional in each getter method (`length(&self, type_mode: TypeMode) -> Option<u16>`, etc) that checked if `self.type_name == "string"`.
  - Rejected since this is a mildly hacky, requiring a conditional check for all types.
- Making `BsonTypeInfo` a trait and making each BSON type its own implementation of the trait, so we could target `String` more precisely.
  - Rejected since this requires a pretty substantial rewrite.
- Mocking the dynamic dispatch we'd achieve from the previous idea without actually going so far as to make `BsonTypeInfo` a trait.
  - This is what I went with by turning these fields into methods. It cleanly allows us to dispatch the handling of `max_string_length` to each type so `String` can be precisely targeted without needing to do a conditional check every time.